sympy/matrices/matrices.py

```diff
@@ -2402,21 +2435,21 @@ def dot(self, b):
                     type(b))
 
         mat = self
-        if mat.cols == b.rows:
-            if b.cols != 1:
-                mat = mat.T
-                b = b.T
-            prod = flatten((mat * b).tolist())
-            if len(prod) == 1:
-                return prod[0]
-            return prod
-        if mat.cols == b.cols:
-            return mat.dot(b.T)
-        elif mat.rows == b.rows:
-            return mat.T.dot(b)
-        else:
-            raise ShapeError("Dimensions incorrect for dot product: %s, %s" % (
-                self.shape, b.shape))
+        if (1 not in mat.shape) or (1 not in b.shape) :
+            SymPyDeprecationWarning(
+                feature="Old feature",
+                issue=13815,
+                deprecated_since_version="1.2").warn()
+            return mat._legacy_array_dot(b)
+        if len(mat) != len(b):
+            raise ShapeError("Dimensions incorrect for dot product: %s, %s" % (self.shape, b.shape))
+        n = len(mat)
+        if mat.shape != (1, n):
```

Comment:
rather than going through the reshaping -- and assuming you want this low-level function to be fast -- would it be better to do `return sum(i*j for i, j in zip(mat._mat, b))` where `b = b._mat if isinstance(b, Matrix) else b`?

Reply:
@smichr reshaping should be very fast, since the internal structure of matrices is flat, it only involves changing a number. Also, we should avoid any references to `_mat` in this code. `_mat` is only appropriate to reference inside the `DenseMatrix` object.